File: Named Pipe Example C#/frmMain.cs
using Named_Pipe_Example.modX;
using System;
using System.Diagnostics;
using System.Threading.Tasks;
using System.Windows.Forms;
using static Named_Pipe_Example.modX.modGlobal;

namespace Named_Pipe_Example
{
    public partial class frmMain : Form
    {
        private clsNamedPipe pipeHelper;
        private System.Windows.Forms.Timer timerProcess;
        private System.Windows.Forms.Timer DoProcess;
        private string chkType = QType.qInbound;

        public frmMain()
        {
            InitializeComponent();
            this.Text = modGlobal.confDevTitleName;
        }

        private int iStatus = 1;
        private int iStep = 0;
        private int ReceivediStatus;
        private int ReceivediStep;

        private bool _isSending = false;

        private async void TimerNamedPipe_Tick(object sender, EventArgs e)
        {
            try
            {
                if (pipeHelper == null || _isSending) return;

                tsslPipeStatus_Value.Text = pipeHelper.IsConnected ? "Connected" : "Not Connected";

                if (modGlobal.confDevDualMain == 1 && pipeHelper.IsConnected)
                {
                    var data = new PipeData { pStatus = iStatus, pStep = iStep };

                    try
                    {
                        _isSending = true;
                        bool sent = await pipeHelper.SendAsync(data);
                        if (!sent)
                        {
                            Debug.WriteLine("Data not sent: pipe busy or disconnected.");
                        }
                    }
                    catch (Exception ex)
                    {
                        Debug.WriteLine($"SendAsync error: {ex.Message}");
                    }
                    finally
                    {
                        _isSending = false;
                    }
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"TimerNamedPipe_Tick error: {ex.Message}");
            }
        }

        private async void Form1_Load(object sender, EventArgs e)
        {
            if (modGlobal.confDevDualMain == 1)
                pipeHelper = new clsNamedPipe(modGlobal.confDevPipeName, isServer: true);
            else
                pipeHelper = new clsNamedPipe(modGlobal.confDevPipeName, isServer: false);

            pipeHelper.OnDataReceived += PipeHelper_OnDataReceived;

            _ = pipeHelper.StartAsync();

            timerProcess = new System.Windows.Forms.Timer();
            timerProcess.Interval = modGlobal.confTimerProcess;
            timerProcess.Tick += TimerNamedPipe_Tick;
            timerProcess.Start();

            DoProcess = new System.Windows.Forms.Timer();
            DoProcess.Interval = modGlobal.confTimerProcess;
            DoProcess.Tick += DoProcess_Tick;
            DoProcess.Start();

            lblCurrentStatus_Value.Text = iStatus.ToString();
            lblCurrentStep_Value.Text = iStep.ToString();
        }

        private async void SetCase(int step)
        {
            iStep = step;
            lblCurrentStep_Value.Text = iStep.ToString();
        }

        private void btnStart_Click(object sender, EventArgs e)
        {
            iStatus = 98;
            iStep = 1;
            lblCurrentStatus_Value.Text = iStatus.ToString();
            lblCurrentStep_Value.Text = iStep.ToString();
        }

        private async void DoProcess_Tick(object sender, EventArgs e)
        {
            try
            {
                switch (chkType)
                {
                    case QType.qInbound:
                        switch (iStep)
                        {
                            case 1:
                                await Task.Delay(900);
                                SetCase(step: 2);
                                break;
                            case 2:
                                await Task.Delay(900);
                                SetCase(step: 3);
                                break;
                            case 3:
                                await Task.Delay(900);
                                SetCase(step: 4);
                                break;
                            case 4:
                                await Task.Delay(900);
                                SetCase(step: 5);
                                break;
                            case 5:
                                await Task.Delay(900);
                                SetCase(step: 6);
                                break;
                            case 6:
                                if (modGlobal.confDevDualMain == 2 && pipeHelper.IsConnected && ReceivediStatus != 1 && ReceivediStep <= 8)
                                {
                                    // รอจนกว่า G02 จะ Step > 8
                                    return;
                                }
                                else
                                {
                                    await Task.Delay(900);
                                    SetCase(step: 7);
                                }
                                break;
                            case 7:
                                await Task.Delay(900);
                                SetCase(step: 8);
                                break;
                            case 8:
                                await Task.Delay(900);
                                SetCase(step: 9);
                                break;
                            case 9:
                                await Task.Delay(900);
                                SetCase(step: 10);
                                break;
                            case 10:
                                await Task.Delay(900);
                                SetCase(step: 11);
                                break;
                            case 11:
                                await Task.Delay(900);
                                SetCase(step: 12);
                                break;
                            case 12:
                                await Task.Delay(900);
                                SetCase(step: 13);
                                break;
                            case 13:
                                await Task.Delay(900);
                                SetCase(step: 0);
                                iStatus = 1;
                                lblCurrentStatus_Value.Text = iStatus.ToString();
                                break;
                        }
                        break;
                    // case อื่นๆ ของ chkType
                    default:
                        break;
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"DoProcess_Tick error: {ex.Message}");
            }
        }

        private void PipeHelper_OnDataReceived(PipeData data)
        {
            // ตัวอย่าง: อัพเดต UI ต้อง invoke เพราะมาจาก thread อื่น
            if (InvokeRequired)
            {
                Invoke(new Action(() => PipeHelper_OnDataReceived(data)));
                return;
            }

            if (pipeHelper.IsConnected)
            {
                ReceivediStatus = data.pStatus;
                ReceivediStep = data.pStep;

                lblPipeStatus_Value.Text = data.pStatus.ToString();
                lblPipeStep_Value.Text = data.pStep.ToString();
            }
        }

        private void Form1_FormClosing(object sender, FormClosingEventArgs e)
        {
            pipeHelper?.Dispose();
        }
    }

}

File: Named Pipe Example Vb/frmMain.vb
Imports Named_Pipe_Example_Vb.Named_Pipe_Example

Public Class frmMain

    Private pipeHelper As NamedPipeHelper
    Private timerProcess As System.Windows.Forms.Timer
    Private DoProcess As System.Windows.Forms.Timer
    Private chkType As String = QType.qInbound

    'Constructor
    Public Sub New()
        ' โค้ดที่ต้องการให้ทำงานตอนสร้างอ็อบเจกต์
        InitializeComponent() ' ตัวนี้ต้องเรียกเพื่อโหลด UI ของฟอร์ม
        Me.Text = modGlobal.confDevTitleName
    End Sub

    Private Sub frmMain_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        If modGlobal.confDevDualMain = 1 Then
            pipeHelper = New NamedPipeHelper(modGlobal.confDevPipeName, isServer:=True)
        Else
            pipeHelper = New NamedPipeHelper(modGlobal.confDevPipeName, isServer:=False)
        End If

        AddHandler pipeHelper.OnDataReceived, AddressOf PipeHelper_OnDataReceived

        Dim startTask As Task = pipeHelper.StartAsync()

        timerProcess = New System.Windows.Forms.Timer()
        timerProcess.Interval = modGlobal.confTimerProcess
        AddHandler timerProcess.Tick, AddressOf TimerNamedPipe_Tick
        timerProcess.Start()

        DoProcess = New System.Windows.Forms.Timer()
        DoProcess.Interval = modGlobal.confTimerProcess
        AddHandler DoProcess.Tick, AddressOf DoProcess_Tick
        DoProcess.Start()

        lblCurrentStatus_Value.Text = iStatus.ToString()
        lblCurrentStep_Value.Text = iStep.ToString()
    End Sub

    Private iStatus As Integer = 1
    Private iStep As Integer = 0
    Private ReceivediStatus As Integer
    Private ReceivediStep As Integer

    Private _isSending As Boolean = False

    Private Sub PipeHelper_OnDataReceived(data As PipeData)
        ' อัพเดต UI ต้อง Invoke เพราะมาจาก thread อื่น
        If Me.InvokeRequired Then
            Me.Invoke(New Action(Sub() PipeHelper_OnDataReceived(data)))
            Return
        End If

        If pipeHelper IsNot Nothing AndAlso pipeHelper.IsConnected Then
            ReceivediStatus = data.pStatus
            ReceivediStep = data.pStep

            lblPipeStatus_Value.Text = data.pStatus.ToString()
            lblPipeStep_Value.Text = data.pStep.ToString()
        End If
    End Sub

    Private Async Sub TimerNamedPipe_Tick(sender As Object, e As EventArgs)
        Try
            tsslPipeStatus_Value.Text = If(pipeHelper IsNot Nothing AndAlso pipeHelper.IsConnected, "Connected", "Not Connected")

            If pipeHelper Is Nothing OrElse _isSending Then Return

            If modGlobal.confDevDualMain = 1 Then
                Dim data As New PipeData With {
                .pStatus = iStatus,
                .pStep = iStep
            }

                Try
                    _isSending = True
                    Dim sent As Boolean = Await pipeHelper.SendAsync(data)
                    If Not sent Then
                        Debug.WriteLine("Data not sent: pipe busy or disconnected.")
                    End If
                Catch ex As Exception
                    Debug.WriteLine($"SendAsync error: {ex.Message}")
                Finally
                    _isSending = False
                End Try
            End If
        Catch ex As Exception
            Debug.WriteLine($"TimerNamedPipe_Tick error: {ex.Message}")
        End Try
    End Sub

    Private Async Sub DoProcess_Tick(sender As Object, e As EventArgs)
        Try
            Select Case chkType
                Case QType.qInbound
                    Select Case iStep
                        Case 1
                            Await Task.Delay(900)
                            SetCase([step]:=2)
                        Case 2
                            Await Task.Delay(900)
                            SetCase([step]:=3)
                        Case 3
                            Await Task.Delay(900)
                            SetCase([step]:=4)
                        Case 4
                            Await Task.Delay(900)
                            SetCase([step]:=5)
                        Case 5
                            Await Task.Delay(900)
                            SetCase([step]:=6)
                        Case 6
                            If modGlobal.confDevDualMain = 2 AndAlso pipeHelper IsNot Nothing AndAlso pipeHelper.IsConnected AndAlso ReceivediStatus <> 1 AndAlso ReceivediStep <= 8 Then
                                ' รอจนกว่า G02 จะ Step > 8
                                Return
                            Else
                                Await Task.Delay(900)
                                SetCase([step]:=7)
                            End If
                        Case 7
                            Await Task.Delay(900)
                            SetCase([step]:=8)
                        Case 8
                            Await Task.Delay(900)
                            SetCase([step]:=9)
                        Case 9
                            Await Task.Delay(900)
                            SetCase([step]:=10)
                        Case 10
                            Await Task.Delay(900)
                            SetCase([step]:=11)
                        Case 11
                            Await Task.Delay(900)
                            SetCase([step]:=12)
                        Case 12
                            Await Task.Delay(900)
                            SetCase([step]:=13)
                        Case 13
                            Await Task.Delay(900)
                            SetCase([step]:=0)
                            iStatus = 1
                            lblCurrentStatus_Value.Text = iStatus.ToString()
                    End Select
                Case Else
                    ' กรณีอื่นของ chkType
            End Select
        Catch ex As Exception
            Debug.WriteLine($"DoProcess_Tick error: {ex.Message}")
        End Try
    End Sub

    Private Sub btnStart_Click(sender As Object, e As EventArgs) Handles btnStart.Click
        iStatus = 98
        iStep = 1
        lblCurrentStatus_Value.Text = iStatus.ToString()
        lblCurrentStep_Value.Text = iStep.ToString()
    End Sub

    Private Sub SetCase([step] As Integer)
        iStep = [step]
        lblCurrentStep_Value.Text = iStep.ToString()
    End Sub

    Private Sub frmMain_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing
        pipeHelper?.Dispose()
    End Sub
End Class
